### PR TITLE
Feature/password reset

### PR DESCRIPTION
**Pull request by dabernathy89**
adds ability to reset password with the existing login shortcode. this is borrowed heavily from wp-login.php.

### CODE REVIEW

**Comment by tnorthcutt** [comment]
@pippinsplugins ping 😉

**Comment by pippinsplugins** [comment]
Awesome! I'm going to leave a few inline comments with minor tweaks.

File: includes/login-functions.php [hunk 1]
```diff
@@ -124,4 +124,161 @@ function rcp_process_login_form() {
...
 }
-add_action('init', 'rcp_process_login_form');
\ No newline at end of file
+add_action('init', 'rcp_process_login_form');
+
+/**
+ * Process the password reset. adapted from wp-login.php
+ *
+ * @access      public
+ * @since       1.0
```
Comment (pippinsplugins) [inline includes/login-functions.php]: Since should be 2.3

File: includes/login-functions.php [hunk 2]
```diff
@@ -124,4 +124,161 @@ function rcp_process_login_form() {
...
+
+add_action('init', 'rcp_process_lostpassword_reset');
+
+// lostpassword_reset
+
+/**
+ * Process the lost password form
+ *
+ * @access      public
+ * @since       1.0
```
Comment (pippinsplugins) [inline includes/login-functions.php]: 2.3

File: includes/login-functions.php [hunk 3]
```diff
@@ -124,4 +124,161 @@ function rcp_process_login_form() {
...
+		exit();
+	}
+}
+add_action('init', 'rcp_process_lostpassword_form');
+
+/**
+ * Send password reset email to user. Adapted from wp-login.php
+ *
+ * @access      public
+ * @since       1.0
```
Comment (pippinsplugins) [inline includes/login-functions.php]: 2.3

File: includes/login-functions.php [hunk 4]
```diff
@@ -124,4 +124,161 @@ function rcp_process_login_form() {
...
+ */
+function rcp_retrieve_password() {
+	global $wpdb, $wp_hasher, $wp_db_version;
+
+	if ( empty( $_POST['rcp_user_login'] ) ) {
+		rcp_errors()->add( 'empty_username', __( 'Enter a username or e-mail address.', 'rcp' ), 'lostpassword' );
+	} elseif ( strpos( $_POST['rcp_user_login'], '@' ) ) {
+		$user_data = get_user_by( 'email', trim( $_POST['rcp_user_login'] ) );
+		if ( empty( $user_data ) )
```
Comment (pippinsplugins) [inline includes/login-functions.php]: Please add brackets around the conditional.

File: includes/login-functions.php [hunk 5]
```diff
@@ -124,4 +124,161 @@ function rcp_process_login_form() {
...
+			rcp_errors()->add( 'invalid_email', __( 'There is no user registered with that email address.', 'rcp' ), 'lostpassword' );
+	} else {
+		$login = trim($_POST['rcp_user_login']);
+		$user_data = get_user_by('login', $login);
+	}
+
+	if ( rcp_errors()->get_error_code() )
```
Comment (pippinsplugins) [inline includes/login-functions.php]: Please add brackets

File: includes/login-functions.php [hunk 6]
```diff
@@ -124,4 +124,161 @@ function rcp_process_login_form() {
...
+	$wpdb->update( $wpdb->users, array( 'user_activation_key' => $hashed ), array( 'user_login' => $user_login ) );
+
+	$message = __('Someone requested that the password be reset for the following account:') . "\r\n\r\n";
+	$message .= network_home_url( '/' ) . "\r\n\r\n";
+	$message .= sprintf(__('Username: %s'), $user_login) . "\r\n\r\n";
+	$message .= __('If this was a mistake, just ignore this email and nothing will happen.') . "\r\n\r\n";
+	$message .= __('To reset your password, visit the following address:') . "\r\n\r\n";
+	$message .= '<' . esc_url($_POST['rcp_redirect']) . "?rcp_action=lostpassword_reset&key=$key&login=" . rawurlencode($user_login) . ">\r\n";
+
+	if ( is_multisite() )
```
Comment (pippinsplugins) [inline includes/login-functions.php]: Please add brackets

File: includes/login-functions.php [hunk 7]
```diff
@@ -124,4 +124,161 @@ function rcp_process_login_form() {
...
+		 */
+		$blogname = wp_specialchars_decode(get_option('blogname'), ENT_QUOTES);
+
+	$title = sprintf( __('[%s] Password Reset'), $blogname );
+
+	$title = apply_filters( 'retrieve_password_title', $title );
+
+	$message = apply_filters( 'retrieve_password_message', $message, $key, $user_login, $user_data );
+
+	if ( $message && !wp_mail( $user_email, wp_specialchars_decode( $title ), $message ) )
```
Comment (pippinsplugins) [inline includes/login-functions.php]: Please add brackets

File: includes/member-forms.php [hunk 8]
```diff
@@ -27,6 +27,30 @@ function rcp_login_form_fields( $args = array() ) {
 	return ob_get_clean();
 }
 
+// lost password form fields
```
Comment (pippinsplugins) [inline includes/member-forms.php]: Please add the appropriate PHPDoc here and on rcp_lostpassword_checkemail_message()

File: includes/member-functions.php [hunk 9]
```diff
@@ -825,6 +833,19 @@ function rcp_change_password() {
 }
 add_action( 'init', 'rcp_change_password' );
 
+function rcp_get_user_resetting_password($rp_path, $rp_cookie) {
```
Comment (pippinsplugins) [inline includes/member-functions.php]: Please add appropriate PHPDoc

File: templates/login.php [hunk 10]
```diff
@@ -21,7 +21,7 @@
 				<label for="rcp_user_remember"><?php _e( 'Remember', 'rcp' ); ?></label>
 				<input type="checkbox" name="rcp_user_remember" id="rcp_user_remember" value="1"/>
 			</p>
-			<p class="rcp_lost_password"><a href="<?php echo esc_url( wp_lostpassword_url( rcp_get_current_url() ) ); ?>"><?php _e( 'Lost your password?', 'rcp' ); ?></a></p>
+			<p class="rcp_lost_password"><a href="<?php echo esc_url( rcp_get_current_url() ); ?>?rcp_action=lostpassword"><?php _e( 'Lost your password?', 'rcp' ); ?></a></p>
```
Comment (pippinsplugins) [inline templates/login.php]: This should not assume the URL doesn't have any query args already. Use `add_query_arg` to add the action

File: templates/lostpassword.php [hunk 11]
```diff
@@ -0,0 +1,29 @@
+<?php if( ! is_user_logged_in() ) : ?>
+
+    <?php rcp_show_error_messages( 'lostpassword' ); ?>
+
+    <form id="rcp_lostpassword_form" class="rcp_form" method="POST" action="<?php echo esc_url( rcp_get_current_url() ); ?>?rcp_action=lostpassword">
```
Comment (pippinsplugins) [inline templates/lostpassword.php]: This should not assume the URL doesn't have any query args already. Use add_query_arg to add the action

**Comment by tnorthcutt** [comment]
@dabernathy89 do you mind addressing these?

**Comment by dabernathy89** [comment]
@tnorthcutt Definitely! I'm in Peru until Sunday morning, but I can take care of these by Monday evening.

**Comment by tnorthcutt** [comment]
@dabernathy89 awesome, thanks!

**Comment by dabernathy89** [comment]
@tnorthcutt @pippinsplugins got slightly delayed - but I think this takes care of all your notes. let me know if you find any other issues!

**Comment by pippinsplugins** [comment]
@dabernathy89 Excellent, testing it now.

**Comment by pippinsplugins** [comment]
@dabernathy89 The form displays great but the email is missing the reset link: https://cloudup.com/c7dk0QI8t1N

File: includes/login-functions.php [hunk 12]
```diff
@@ -124,4 +124,163 @@ function rcp_process_login_form() {
...
+	}
+
+	$wpdb->update( $wpdb->users, array( 'user_activation_key' => $hashed ), array( 'user_login' => $user_login ) );
+
+	$message = __('Someone requested that the password be reset for the following account:') . "\r\n\r\n";
+	$message .= network_home_url( '/' ) . "\r\n\r\n";
+	$message .= sprintf(__('Username: %s'), $user_login) . "\r\n\r\n";
+	$message .= __('If this was a mistake, just ignore this email and nothing will happen.') . "\r\n\r\n";
+	$message .= __('To reset your password, visit the following address:') . "\r\n\r\n";
+	$message .= '<' . esc_url($_POST['rcp_redirect']) . "?rcp_action=lostpassword_reset&key=$key&login=" . rawurlencode($user_login) . ">\r\n";
```
Comment (pippinsplugins) [inline includes/login-functions.php]: This should use add_query_arg(). It's possible for the redirect to already have query args attached to it.

**Comment by pippinsplugins** [comment]
Let's also please put the new password functions in login-functions.php instead of member-functions.php

**Comment by dabernathy89** [comment]
@pippinsplugins could you see if the URL is still not showing in your email? seems to be working on my end.

Also, is esc_url_raw safe to use there? esc_url() encodes the ampersands, which breaks the link.

**Comment by tnorthcutt** [comment]
@pippinsplugins ping :wink:

**Comment by pippinsplugins** [comment]
Testing this now.

**Comment by pippinsplugins** [comment]
I still don't see the URL: https://cloudup.com/c_LLbQXQ4UZ

I'm using the wpMandrill plugin.

**Comment by tnorthcutt** [comment]
@dabernathy89 ping^ – let me know if I need to jump in and take a look

**Comment by pippinsplugins** [comment]
Found! Checking this fixes it: https://cloudup.com/cjWt1yXDQbI

**Comment by dabernathy89** [comment]
Whew! I wondered if it might be Mandrill seeing those brackets and doing something weird with them, guess it's not that. Do you know if this affected the standard WordPress email reset email too?

**Comment by pippinsplugins** [comment]
It probably does affect standard WP ones.

**Comment by pippinsplugins** [comment]
Awesome work @dabernathy89 @tnorthcutt !

**Comment by tnorthcutt** [comment]
fresh-prince-happy-dance

**Pull request merged**